uniffi_bindgen/src/bindings/gecko/gen_gecko.rs

```diff
@@ -0,0 +1,355 @@
+            Type::UInt64 => "uint64_t".into(),
+            Type::Float32 => "float".into(),
+            Type::Float64 => "double".into(),
+            Type::Boolean => "bool".into(),
+            Type::String => "nsAString".into(),
+            Type::Enum(name) | Type::Record(name) => class_name_cpp(name)?,
+            Type::Object(name) => format!("RefPtr<{}>", class_name_cpp(name)?),
+            Type::Error(name) => panic!("[TODO: type_cpp({:?})]", type_),
+            Type::Optional(type_) => format!("Nullable<{}>", type_cpp(type_)?),
+            Type::Sequence(type_) => format!("nsTArray<{}>", type_cpp(type_)?),
+            Type::Map(type_) => format!("Record<nsString, {}>", type_cpp(type_)?),
+        })
+    }
+
+    /// Declares the type of a return value from C++.
+    pub fn ret_type_cpp(type_: &Type) -> Result<String, askama::Error> {
+        Ok(match type_ {
+            Type::Int8
+            | Type::UInt8
+            | Type::Int16
+            | Type::UInt16
+            | Type::Int32
+            | Type::UInt32
+            | Type::Int64
+            | Type::UInt64
+            | Type::Float32
+            | Type::Float64
+            | Type::Boolean
+            | Type::Enum(_) => type_cpp(type_)?,
+            Type::String => "nsAString&".into(),
+            Type::Object(name) => format!("already_AddRefed<{}>", class_name_cpp(name)?),
+            Type::Error(name) => panic!("[TODO: ret_type_cpp({:?})]", type_),
+            Type::Record(_) | Type::Optional(_) | Type::Sequence(_) | Type::Map(_) => {
+                format!("{}&", type_cpp(type_)?)
+            }
+        })
+    }
+
+    pub fn var_name_webidl(nm: &dyn fmt::Display) -> Result<String, askama::Error> {
+        Ok(nm.to_string().to_mixed_case())
+    }
+
+    pub fn enum_variant_webidl(nm: &dyn fmt::Display) -> Result<String, askama::Error> {
+        Ok(nm.to_string().to_mixed_case())
+    }
+
+    pub fn class_name_webidl(nm: &dyn fmt::Display) -> Result<String, askama::Error> {
+        Ok(nm.to_string().to_camel_case())
+    }
+
+    pub fn class_name_cpp(nm: &dyn fmt::Display) -> Result<String, askama::Error> {
+        Ok(nm.to_string().to_camel_case())
+    }
+
+    pub fn fn_name_webidl(nm: &dyn fmt::Display) -> Result<String, askama::Error> {
+        Ok(nm.to_string().to_mixed_case())
+    }
+
+    /// For interface implementations, function and methods names are
+    /// UpperCamelCase, even though they're mixedCamelCase in WebIDL.
+    pub fn fn_name_cpp(nm: &dyn fmt::Display) -> Result<String, askama::Error> {
+        Ok(nm.to_string().to_camel_case())
+    }
+
+    pub fn field_name_cpp(nm: &str) -> Result<String, askama::Error> {
+        Ok(format!("m{}", nm.to_camel_case()))
+    }
+
+    pub fn enum_variant_cpp(nm: &dyn fmt::Display) -> Result<String, askama::Error> {
+        // TODO: Make sure this does the right thing for hyphenated variants.
+        // Example: "bookmark-added" becomes `Bookmark_added`.
```

Comment:
Let’s add a test case for hyphenated enum variants to UniFFI, and verify it gets reflected correctly in the binding layer.